### ISSUE TITLE
New entries to dataset.dict.csv

### ISSUE THREAD

**Issue by hsellik** [tech-srl/code2vec#63]
Hi,

I am currently trying to preprocess `java-large` on a server with 2 x Intel Xeon E5-2660 v3 which in total has 20 cores and 40 threads and 378GB of RAM. However, it is taking very long as mentioned in https://github.com/tech-srl/code2vec/issues/3. Currently, I've been running it for 12 hours and I am still extracting paths for training set. I have a few questions regarding this.

1) Is there a way to update the resulting `dataset_name.dict.csv` file with only one project? The use case is the following. When I want to add new project(s) to train/test/validation set, then instead of preprocessing the whole dataset again, I'd like to append them to the `dataset_name.dict.csv` 

1.1) I presume `dict.csv` file is important when training from scratch, not so important when continuing training since there is no need for the old dict.csv entries anyway?

2) What is the purpose of the `JavaExtractor/extract.py`? I am asking because calling many jar instances seems to take quite a while and `JavaExtractor-0.0.1-SNAPSHOT.jar` creates threads internally for every file anyway if called with `--dir` argument. I'd call it directly and make some enhancements inside to speed it up (skipping extract.py). Is there stuff that I might miss by doing this? The `raw.txt` files seem to only contain output from `jar` and nothing else.

Hope my questions do make sense.

Thanks in advance,
Hendrig

**Comment by urialon** [comment]
Hi Hendrig,
Yes, unfortunately processing such large datasets takes some time. 
You can decrease the timeout [here](https://github.com/tech-srl/code2vec/blob/master/JavaExtractor/extract.py#L39), to 5 minutes per-directory or so, such that directories that are not completed in time are skipped and their subdirs are extracted recursively (except the few problematic subdirs which will be skipped). 
This will be your best solution, I think.

Another option is to try java-med/small first?

As for your questions:
1. Yes, it is possible. The `dict.c2v` file is a pickle file of a few dictionaries. It is created [here](https://github.com/tech-srl/code2vec/blob/master/preprocess.py#L16). So if you load an existing dict, update its values, and re-write it with the new values it is supposed to be fine (but not continue training with a model that has started training with the old `dict.c2v`! only to start training a new model.

1.1 This is correct, when loading a trained model, the necessary dictionaries are loaded from the model's files, not from the dataset's files. This is done [here](https://github.com/tech-srl/code2vec/blob/master/vocabularies.py#L178).

2. The purpose of `extract.py` is to wrap `JavaExtractor-0.0.1-SNAPSHOT.jar` by running it using multiple processes, while each jar uses multiple threads... This hacky structure worked faster on my machines because sometimes too large files take all the JVM memory and stuck the whole `jar` running process.
Additionally, if a `jar` process is taking too much time - it is killed from outside (from the `extract.py` script) and then re-ran recursively on its subdirs. 
Skipping `extract.py` is possible, but I don't see how it will speed up anything. If you wish to run only a single process at a time, you can just set it [here](https://github.com/tech-srl/code2vec/blob/master/JavaExtractor/extract.py#L68) (replace `4` with `1`).

I hope it helps!

**Comment by hsellik** [comment]
Okay, I see, it makes more sense now.

If I want to add the additional data to initial dataset, I'll update the pickle file so others can train it from scratch instead of providing 2 datasets which people could also theoretically train on sequentially.

However, I do have one question. 

1) If I want to keep training the model, I'll just create a new entirely separate dataset with no pickle editing and `use the --load flag to load the trained model; the --data flag to point to the new dataset to train on; and the --save flag to provide a new save path`. This case is different and I presume it is okay to keep training a model that started with the old `dict.c2v` ?

**Comment by urialon** [comment]
Hi @hsellik ,
So if I understand correctly, you wish to do transfer learning where you train on dataset-A, and then continue training or fine-tune on dataset-B? 
This is definitely possible, just keep in mind that if the first training set was not large+diverse enough, some tokens and paths in the second training set will be treated as "unknown".

Alternatively, if you add names and paths from the new dataset ("Dataset-B") into the existing dictionaries this may work as well. The only problem is that "old names" will be already trained, and mixed with "new names" that appeared first in the training set of dataset-B (which will start as randomly initialized vectors), but I guess there's no other choice in this case.

I am guessing that [code2seq](https://github.com/tech-srl/code2seq/) will perform much much better at this transfer task, because code2vec is much more sparse.

But that's a cool idea, I would love to hear how it went.

**Comment by hsellik** [comment]
Hi, @urialon ,

Yes, I want to keep training / further train the model. Although I don't know if it can be classified as "transfer learning" since the problem remains the same.

In summary, I have trained a binary classification model on a dataset  ("Dataset-A") and want to further train on a new dataset ("Dataset-B") with the same problem and no new labels, but avoid tokens and paths being treated as "unknown". 

I want to do that because I have a fair amount of computing power for Open Source Software projects, but very limited computing capability at the location of the target project. The classification problem remains the same.

Basically, the question is: do the [Further training instructions in README](https://github.com/tech-srl/code2vec#downloading-a-trained-model-35-gb-which-can-be-further-trained) suffer the same "unknown token" problem as described in our discussion?

**Comment by urialon** [comment]
Hi @hsellik ,
OK I see.

> Basically, the question is: do the Further training instructions in README suffer the same "unknown token" problem as described in our discussion?
Yes. It does suffer. 

BUT, I think that if Dataset-A is large enough, you'll find that most (90%?) of the tokens and paths in Dataset-B are already in the vocabulary. Maybe the easiest first step will be checking that hypothesis by simply counting / comparing histograms, before training? You can check both:

1. How many of the unique paths/tokens from Dataset-B are already in the vocabulary created by preprocessing Dataset-A
2. How many of the **occurrences** of paths/tokens  from Dataset-B are already in the vocabulary created by preprocessing Dataset-A (I guess that this one will be much higher than the former)

Then you can try both:
1. Continue training on Dataset-B, without any change to the vocabularies.
2. Continue training on Dataset-B, but add Dataset-B-specific common tokens+paths into the dictionaries and into the embedding matrices (will require a minor tensorflow work of taking the already-trained embedding matrix, and extending it with fresh, uninitialized rows).

I am also guessing that if you work with code2seq instead, you won't need any vocabulary-extensions or modifications and you will be able to just continue training on Dataset-B (since code2seq works with nodes instead of full paths, you won't have unknown nodes in Dataset-B; since code2seq works with subtokens instead of tokens, the probability of observing an unknown *sub*token is very small)

Good luck, let me know how it goes!

**Comment by hsellik** [comment]
That's perfect, thank you for the very thorough answers! I'll let you know how it goes. Will definitely use code2seq as well and compare the results.
